File: leetcode/cpp/two_pointers/lc_15_3sum.cpp
#include <vector>
#include <algorithm>

/**
 * https://leetcode.com/problems/3sum/description/
 *
 * Time complexity: O(n^2) where n = length of nums
 * Space complexity: O(n^2) where n = length of nums
 */
class Solution {
public:
    std::vector<std::vector<int>> threeSum(std::vector<int>& nums) {
        // first sort the nums array
        // this will allows us to use two pointer sorted 2sum technique
        // and place duplicates right next to each other
        // so we can skip the duplicates easily
        std::sort(nums.begin(), nums.end());

        // init our results array
        std::vector<std::vector<int>> result;

        // loop through all the nums indices except for the last two
        // this index i will point to our first number
        for (int i = 0; i < nums.size()-2; ++i) {
            // skip duplicates for the first number at index i
            if (i > 0 && nums[i-1] == nums[i]) {
                continue;
            }

            // setup our left and right pointers
            // as we did for sorted 2sum technique
            int left = i+1;
            int right = nums.size()-1;

            // standard while loop as long as the two pointers don't overlap
            while (left < right) {
                // compute the sum of the three numbers
                int sum = nums[i] + nums[left] + nums[right];

                // if the sum is 0, we have a valid combination
                if (sum == 0) {
                    // create our triplet vector and add all numbers to it
                    std::vector<int> triplet;
                    triplet.push_back(nums[i]);
                    triplet.push_back(nums[left]);
                    triplet.push_back(nums[right]);
                    result.push_back(triplet);

                    // update left and right pointers
                    // because we want to try out new combinations now
                    ++left;
                    --right;

                    // skip duplicates for the second number at the left pointer
                    // we don't need to skip duplicates for the right pointer
                    // because as long as 2 out of 3 numbers are unique, the triplet is unique
                    while (left < right && nums[left-1] == nums[left]) {
                        ++left;
                    }

                }
                // else check if our sum is too big
                // so decrease the right pointer and make it smaller
                else if (sum > 0) {
                    --right;
                }
                // else our sum is too small
                // so increase the left pointer and make it larger
                else {
                    ++left;
                }
            }
        }

        // finally return the result vector
        return result;
    }
};

File: leetcode/two_pointers/lc_15_3sum.py
from typing import List


'''
https://leetcode.com/problems/3sum/description/
'''
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # first we will sort the array
        # to place duplicates next to one another
        nums.sort()

        # init our result array
        result = []

        # loop through all indices in nums
        for i in range(len(nums)):
            # if we find any duplicate numbers
            # after the first number at i = 0
            if i > 0 and nums[i-1] == nums[i]:
                # then skip it so continue
                continue

            # setup our two pointer approach
            # left pointer points to number after i
            left = i+1
            # right pointer points to last number in array
            right = len(nums)-1

            # loop as long as the left/right pointers don't overlap
            while left < right:
                # compute the three_sum
                three_sum = nums[i] + nums[left] + nums[right]

                # if our three sum = 0 then we have a valid combo
                if three_sum == 0:
                    # add this combo to our results array
                    result.append([nums[i], nums[left], nums[right]])

                    # now we want to try out another combination, so
                    # increment our left pointer
                    left += 1
                    # decrement our right pointer
                    right -= 1

                    # skip duplicates on the left pointer
                    # we don't need to skip duplicates on right pointer
                    # since out of our 3 numbers, as long as 2 out of 3
                    # are unique, we are ok and will never had duplicate entries
                    while left < right and nums[left-1] == nums[left]:
                        # increment our left pointer until we don't have duplicates
                        left += 1

                # else if our three sum is less than 0
                elif three_sum < 0:
                    # then increment the left pointer
                    # to increase the three sum
                    left += 1
                # else our three sum is greater than 0
                else:
                    # the decrement the right pointer
                    # to decrease the three sum
                    right -= 1

        # finally, return the results array
        return result
